_posts/2018-01-01-testing-code-that-isnt-there-yet.md

```diff
@@ -0,0 +1,671 @@
+            const routeHandler = createHotelPhotosRouteHandler(connectedClientDouble, collectionName);
+            const photosCollection = routeHandler(ctxDouble); // (2)
+
+            expect(photosCollection).to.deep.equal([
+                'photo-1.jpg', 'photo-2.jpg', 'photo-3.jpg'
+            ]);
+        });
+    });
+});
+```
+
+```javascript
+export default function createHotelPhotosRouteHandler(dbClient, collectionName) {
+    dbClient.collection(collectionName);
+
+    return () => [ 'photo-1.jpg', 'photo-2.jpg', 'photo-3.jpg' ];
+}
+```
+
+1. It is a very good practice (most of the time a requirement) to reset calls count for spies
+   if we check in tests if they were called given amount of times. Remember that you should
+   be able to call all unit tests independently, at any given moment, in any order. That being
+   said, one spy call should not affect the other call in other test. Same goes for stubs.
+
+1. At this point, we still want to have photos returned. But since this is a route handler,
+   I guess it should not return a plain array, but a response with status code and a body
+   holding that array.
+
+<blockquote>
+For the sake of readability, I will add only a single test here and there to not overwhelme
+anyone with the whole codebase.
+</blockquote>
+
+In order for Koa to return a response with given status code and a body,
+we need to set `status` and `body` properties of `response` property of `ctx` (I will explain it shortly).
+This is, again, a thing worth discovering before doing any coding.<br>
+
+### "Learn how to use your tool, before using it."
+<br>
+So, if we want to set `200` and a body with that collection of photos, we need to do something like:
+
+```javascript
+ctx.response.status = 200;
+ctx.response.body = [ 'photo-1.jpg', 'photo-2.jpg', 'photo-3.jpg' ];
+```
+
+Let's write a test for that:
+
+```javascript
+const ctxDouble = {
+    response: { // (1)
+        status: 0,
+        body: ''
+    }
+};
+
+beforeEach(() => {
+    connectedClientDouble.collection.reset();
+
+    ctxDouble.response.status = 0; // (2)
+    ctxDouble.response.body = '';
+});
+
+it('should return hotel photos collection', () => { // (3)
+    const routeHandler = createHotelPhotosRouteHandler(connectedClientDouble, collectionName);
+
+    routeHandler(ctxDouble); // (4)
+
+    expect(ctxDouble.response.status).to.equal(200);
+    expect(ctxDouble.response.body).to.deep.equal([
+        'photo-1.jpg', 'photo-2.jpg', 'photo-3.jpg'
+    ]);
+});
+```
+
+```javascript
+export default function createHotelPhotosRouteHandler(dbClient, collectionName) {
+    dbClient.collection(collectionName);
+
+    return (ctx) => {
+        ctx.response.status = 200;
+        ctx.response.body = ['photo-1.jpg', 'photo-2.jpg', 'photo-3.jpg'];
+    };
+}
+```
+1. We're adding `response` property which is an object holding `status` and `body` properties,
+   which will be set and checked.
+
+1. They need to be reseted before each test.
```

Comment:
reseted -> reset

Reply:
Hmm ... are you sure?

Reply:
pretty much - see here http://www.verbix.com/webverbix/English/reset.html - it's `reset` all the way down

Reply:
Thanks, nice resource. Was looking one time for a 'dictionary' like this (with all of the forms for a given word etc.).